Compile-time extensions for list-object-filter

olsen232 opened the pull request:
Adds an extension:<custom-filter> option to list-object-filters,
these are implemented by static libraries that can be compiled
into git using the FILTER_EXTENSIONS makefile argument.

No need to actually merge this PR anywhere at this time - this PR is an FYI of where I am at, and a way to comment on any changes.

@rcoup - recent changes:

- plugins / profiles are now named extensions everywhere. (I started naming them filter-profiles in one place, but calling them filter-plugins for addec clarity in another place - to avoid confusion, I just call them extension everywhere, which happily makes sense both to mean "plugin" or "non-standard-functionality".)
- now using a shell script to generate the array of filter extensions, instead of patsubst and preprocessor macros.
- updated the documentation
- got a C++ filter extension to work without any changes to existing git header files. There are some limitations eg you need to be willing to write a little bit of C in addition to C++.
- fixed some bugs I had in the Makefile

Currently I have all the changes in a single commit - I plan to divide the commit up into thematic commits at some point (eg implementation, documentation), but if needed I can rebase it against my last attempt so you can see changes.

olsen232 committed:
Compile-time extensions for list-object-filter

Adds an extension:<custom-filter> option to list-object-filters,
these are implemented by static libraries that can be compiled
into git using the FILTER_EXTENSIONS makefile argument.

rcoup commented on Makefile:
```diff
@@ -1955,6 +1958,20 @@ ifndef PAGER_ENV
+FILTER_EXT_PATHS = $(dir $(FILTER_EXTENSIONS))
+
+$(FILTER_EXTENSIONS): $(FILTER_EXT_PATHS)
+	$(MAKE) -C $(@D) \
```
Shouldn't this use `$(QUIET_SUBDIR)` or something? It's a lot noiser than the other subdirectories when doing a non-verbose build.

olsen232 replied:
Done

rcoup commented on contrib/filter-extensions/README.md:
```diff
@@ -0,0 +1,141 @@
+List-Objects-Filter Extensions API
```
Git uses [asciidoc](https://asciidoc.org/) for documentation 😬

olsen232 replied:
Done

rcoup commented on contrib/filter-extensions/README.md:
```diff
@@ -0,0 +1,141 @@
+C wrapper API that you add to a C++ library that already contains the
+domain-specific operations that you need. In either case, remember to wrap any
+functions that must be C-compatible with `extern C` when declaring or
+defining them from within C++.
```
Reading more, I'm not sure it's quite that simple...

https://isocpp.org/wiki/faq/mixing-c-and-cpp#overview-mixing-langs

> * You must use your C++ compiler when compiling `main()` (e.g., for static initialization)
> * Your C++ compiler should direct the linking process (e.g., so it can get its special libraries)

I think the latter is the same as `-lstdc++`/`-lc++`, but the former is a problem? Not entirely sure of the best approach here. There are some [suggestions](https://stackoverflow.com/questions/66192195/when-mixing-c-and-c-code-does-main-need-to-be-in-the-c-part) that modern-ish clang/gcc (MSVC???) are fine with it.

rcoup replied:
I added a crude runtime initialisation test to `rand.cpp` for testing:

```patch
diff --git a/contrib/filter-extensions/rand_cpp/rand.cpp b/contrib/filter-extensions/rand_cpp/rand.cpp
index a907e999cb..4d27bd85d5 100644
--- a/contrib/filter-extensions/rand_cpp/rand.cpp
+++ b/contrib/filter-extensions/rand_cpp/rand.cpp
@@ -9,6 +9,13 @@ extern "C" {
        #include "adapter_functions.h"
 }

+// run-time dynamic initialisation test
+static int hello() {
+       std::cerr << "filter-rand-cpp: good morning\n";
+       return 123;
+}
+static int A_NUMBER = hello();
+
 struct rand_context {
        int percentageMatch = 0;
        int matchCount = 0;
```

This works (`gcc` == `clang` on macOS by default)
```console
$ make CC=clang LDFLAGS=-lc++ FILTER_EXTENSIONS=contrib/filter-extensions/rand_cpp/rand_cpp.a
$ ./git --version
filter-rand-cpp: good morning
git version 2.32.0.rc1.753.g680de0773f.dirty
$ ./git rev-list refs/heads/master --objects --max-count 1 --filter=extension:rand_cpp=3 --filter-print-omitted | grep -c '^~'
filter-rand-cpp: good morning
filter-rand-cpp: matching 3%
filter-rand-cpp: done: count=4057 (blob=3856 tree=201) matched=117 elapsed=0.008535s rate=475337/s average=2.10377us
3739
```
This doesn't link properly:
```console
$ make CC=gcc-11 LDFLAGS=-lstdc++ FILTER_EXTENSIONS=contrib/filter-extensions/rand_cpp/rand_cpp.a
```

rcoup replied:
Here's a basic patch to enable using an alternate compiler linker driver for linking (eg: `CC_LD=g++` / `CC_LD=clang++` / `CC_LD=g++-11`). 

```patch
diff --git a/Makefile b/Makefile
index 5fbf2de14e..db26a0444d 100644
--- a/Makefile
+++ b/Makefile
@@ -552,6 +552,7 @@ export prefix bindir sharedir sysconfdir gitwebdir perllibdir localedir

 # Set our default programs
 CC = cc
+CC_LD = $(CC)
 AR = ar
 RM = rm -f
 DIFF = diff
@@ -2240,7 +2241,7 @@ git.sp git.s git.o: EXTRA_CPPFLAGS = \
        '-DGIT_INFO_PATH="$(infodir_relative_SQ)"'

 git$X: git.o GIT-LDFLAGS $(BUILTIN_OBJS) $(GITLIBS) $(EXTENSION_LIBS)
-       $(QUIET_LINK)$(CC) $(ALL_CFLAGS) -o $@ $(ALL_LDFLAGS) \
+       $(QUIET_LINK)$(CC_LD) $(ALL_CFLAGS) -o $@ $(ALL_LDFLAGS) \
                $(filter %.o,$^) $(LIBS)

 help.sp help.s help.o: command-list.h
@@ -2604,17 +2605,17 @@ compat/nedmalloc/nedmalloc.sp: SP_EXTRA_FLAGS += -Wno-non-pointer-null
 endif

 git-%$X: %.o GIT-LDFLAGS $(GITLIBS)
-       $(QUIET_LINK)$(CC) $(ALL_CFLAGS) -o $@ $(ALL_LDFLAGS) $(filter %.o,$^) $(LIBS)
+       $(QUIET_LINK)$(CC_LD) $(ALL_CFLAGS) -o $@ $(ALL_LDFLAGS) $(filter %.o,$^) $(LIBS)

 git-imap-send$X: imap-send.o $(IMAP_SEND_BUILDDEPS) GIT-LDFLAGS $(GITLIBS)
-       $(QUIET_LINK)$(CC) $(ALL_CFLAGS) -o $@ $(ALL_LDFLAGS) $(filter %.o,$^) \
+       $(QUIET_LINK)$(CC_LD) $(ALL_CFLAGS) -o $@ $(ALL_LDFLAGS) $(filter %.o,$^) \
                $(IMAP_SEND_LDFLAGS) $(LIBS)

 git-http-fetch$X: http.o http-walker.o http-fetch.o GIT-LDFLAGS $(GITLIBS)
-       $(QUIET_LINK)$(CC) $(ALL_CFLAGS) -o $@ $(ALL_LDFLAGS) $(filter %.o,$^) \
+       $(QUIET_LINK)$(CC_LD) $(ALL_CFLAGS) -o $@ $(ALL_LDFLAGS) $(filter %.o,$^) \
                $(CURL_LIBCURL) $(LIBS)
 git-http-push$X: http.o http-push.o GIT-LDFLAGS $(GITLIBS)
-       $(QUIET_LINK)$(CC) $(ALL_CFLAGS) -o $@ $(ALL_LDFLAGS) $(filter %.o,$^) \
+       $(QUIET_LINK)$(CC_LD) $(ALL_CFLAGS) -o $@ $(ALL_LDFLAGS) $(filter %.o,$^) \
                $(CURL_LIBCURL) $(EXPAT_LIBEXPAT) $(LIBS)

 $(REMOTE_CURL_ALIASES): $(REMOTE_CURL_PRIMARY)
@@ -2624,7 +2625,7 @@ $(REMOTE_CURL_ALIASES): $(REMOTE_CURL_PRIMARY)
        cp $< $@

 $(REMOTE_CURL_PRIMARY): remote-curl.o http.o http-walker.o GIT-LDFLAGS $(GITLIBS)
-       $(QUIET_LINK)$(CC) $(ALL_CFLAGS) -o $@ $(ALL_LDFLAGS) $(filter %.o,$^) \
+       $(QUIET_LINK)$(CC_LD) $(ALL_CFLAGS) -o $@ $(ALL_LDFLAGS) $(filter %.o,$^) \
                $(CURL_LIBCURL) $(EXPAT_LIBEXPAT) $(LIBS)

 $(LIB_FILE): $(LIB_OBJS)
@@ -2930,7 +2931,7 @@ perf: all
 t/helper/test-tool$X: $(patsubst %,t/helper/%,$(TEST_BUILTINS_OBJS))

 t/helper/test-%$X: t/helper/test-%.o GIT-LDFLAGS $(GITLIBS)
-       $(QUIET_LINK)$(CC) $(ALL_CFLAGS) -o $@ $(ALL_LDFLAGS) $(filter %.o,$^) $(filter %.a,$^) $(LIBS)
+       $(QUIET_LINK)$(CC_LD) $(ALL_CFLAGS) -o $@ $(ALL_LDFLAGS) $(filter %.o,$^) $(filter %.a,$^) $(LIBS)

 check-sha1:: t/helper/test-tool$X
        t/helper/test-sha1.sh
```

These both work:
```console
$ make CC=clang CC_LD=clang++ FILTER_EXTENSIONS=contrib/filter-extensions/rand_cpp/rand_cpp.a
$ ./git --version
filter-rand-cpp: good morning
git version 2.32.0.rc1.753.g680de0773f.dirty
$ ./git rev-list refs/heads/master --objects --max-count 1 --filter=extension:rand_cpp=3 --filter-print-omitted | grep -c '^~'
filter-rand-cpp: good morning
filter-rand-cpp: matching 3%
filter-rand-cpp: done: count=4057 (blob=3856 tree=201) matched=117 elapsed=0.010891s rate=372509/s average=2.6845us
3739

$ make CC=gcc-11 CC_LD=g++-11 FILTER_EXTENSIONS=contrib/filter-extensions/rand_cpp/rand_cpp.a
$ ./git --version
filter-rand-cpp: good morning
git version 2.32.0.rc1.753.g680de0773f.dirty
$ ./git rev-list refs/heads/master --objects --max-count 1 --filter=extension:rand_cpp=3 --filter-print-omitted | grep -c '^~'
filter-rand-cpp: good morning
filter-rand-cpp: matching 3%
filter-rand-cpp: done: count=4057 (blob=3856 tree=201) matched=117 elapsed=0.007292s rate=556363/s average=1.79739us
3739
```

rcoup replied:
So I guess need to check it on Linux and under MSVC?

olsen232 replied:
... I wonder how much effort we should go to, to have a C++ filter extension in contrib/ and in the documentation? Mostly, we want to people to be able to a) contribute filters written in C in a straightforward manner or b) use a language of their choice, and use their specific knowledge of that language's C bindings to debug any linking issues. Having a C++ example I suppose is a good way to help people think through the general problem of b), and it's mostly working, so I guess we'll fix it up so it 100% works?

Not yet done, I'll apply your patch in a bit

rcoup replied:
I guess we submit it and see what comments are like, it can always live as an example outside Git's repo if need be.  Feels like if people want to use filters to do Clever Things™ then that could quite often involve some external libraries. If there's a straightforward path to tread that'll certainly help them.

rcoup commented on Makefile:
```diff
@@ -1955,6 +1958,20 @@ ifndef PAGER_ENV
 PAGER_ENV = LESS=FRX LV=-c
 endif
 
+ifneq ($(FILTER_EXTENSIONS),)
```
should add FILTER_EXTENSIONS to the docs at the top of the Makefile too

olsen232 replied:
Done

rcoup commented on contrib/filter-extensions/README.md:
```diff
@@ -0,0 +1,141 @@
+since the filter extensions and Git itself will be linked together into a single
+binary.
+
+You may depend on other libraries if you indicate that they are to be linked
+into the Git binary using `FILTER_EXTENSION_LIBS`. See the C++ example below.
```
Doesn't `FILTER_EXTENSION_LIBS` basically do the same thing as `LDFLAGS=-lwhatever`?

olsen232 replied:
... yerss. Removed FILTER_EXTENSION_LIBS, updated documentation to refer to LDFLAGS.

However, LDFLAGS is set to empty at some point during the Makefile. The Makefile comments refers to this somewhere as a default value that can be overridden, but elsewhere it mentions that setting it to empty makes sure that environment variables don't interfere with it. Setting LDFLAGS in `config.mak` seems to override it but I don't know exactly what the difference is between using environment variables and config.mak.

Also updated documentation to use config.mak whenever LDFLAGS are set.

rcoup replied:
I don't quite understand why, but maybe `LDFLAGS=x make` is different from `make LDFLAGS=x`? The latter was definitely working for me today.

rcoup commented on generate-list-objects-filter-extensions.sh:
```diff
@@ -0,0 +1,42 @@
+#!/bin/sh
```
there's a bunch of shellcheck warnings here:

```
$ shellcheck generate-list-objects-filter-extensions.sh

In generate-list-objects-filter-extensions.sh line 6:
	ARGS=`echo $@ | xargs printf '%s\n'`
             ^-----------------------------^ SC2006: Use $(...) notation instead of legacy backticked `...`.
                   ^-- SC2068: Double quote array expansions to avoid re-splitting elements.

Did you mean:
	ARGS=$(echo $@ | xargs printf '%s\n')


In generate-list-objects-filter-extensions.sh line 9:
	INVALID_ARGS=`echo "$ARGS" | grep -v '\b[A-Za-z0-9_]\+\.a$'`
                     ^-- SC2006: Use $(...) notation instead of legacy backticked `...`.

Did you mean:
	INVALID_ARGS=$(echo "$ARGS" | grep -v '\b[A-Za-z0-9_]\+\.a$')


In generate-list-objects-filter-extensions.sh line 17:
	EXTS=`echo "$ARGS" | sed -e 's!.*/!filter_extension_!' -e 's!.a$!!'`
             ^-- SC2006: Use $(...) notation instead of legacy backticked `...`.

Did you mean:
	EXTS=$(echo "$ARGS" | sed -e 's!.*/!filter_extension_!' -e 's!.a$!!')


In generate-list-objects-filter-extensions.sh line 20:
	DECLARATIONS=`echo "$EXTS" | sed -e 's!^!\t!' -e 's!$!,!'`
                     ^-- SC2006: Use $(...) notation instead of legacy backticked `...`.

Did you mean:
	DECLARATIONS=$(echo "$EXTS" | sed -e 's!^!\t!' -e 's!$!,!')


In generate-list-objects-filter-extensions.sh line 23:
	ARRAY=`echo "$EXTS" | sed -e 's!^!\t\&!' -e 's!$!,!'`
              ^-- SC2006: Use $(...) notation instead of legacy backticked `...`.

Did you mean:
	ARRAY=$(echo "$EXTS" | sed -e 's!^!\t\&!' -e 's!$!,!')


In generate-list-objects-filter-extensions.sh line 41:
echo '\tNULL,'
     ^-------^ SC2028: echo may not expand escape sequences. Use printf.

For more information:
  https://www.shellcheck.net/wiki/SC2068 -- Double quote array expansions to ...
  https://www.shellcheck.net/wiki/SC2028 -- echo may not expand escape sequen...
  https://www.shellcheck.net/wiki/SC2006 -- Use $(...) notation instead of le...
```

olsen232 replied:
Done

rcoup commented on list-objects-filter-extensions.c:
```diff
@@ -0,0 +1,15 @@
+/* Automatically generated by generate-list-objects-filter-extensions.sh */
```
this should be `gitignore`d and not in the source tree?

olsen232 replied:
Done

rcoup commented on contrib/filter-extensions/rand/rand.c:
```diff
@@ -0,0 +1,111 @@
+	ctx->percentageMatch = 1;  // default 1%
+	}
+	fprintf(stderr, "filter-rand: matching %d%%\n", ctx->percentageMatch);
+	clock_gettime(CLOCK_MONOTONIC, &ctx->started_at);
```
should probably use `getnanotime()` from `trace.h` since it's platform-agnostic

olsen232 replied:
Done

rcoup commented on contrib/filter-extensions/rand/rand.c:
```diff
@@ -0,0 +1,111 @@
+
+	switch (filter_situation) {
+	default:
+		fprintf(stderr,
+			"filter-rand: unknown filter_situation: %d\n",
+			filter_situation);
+		abort();
```
`die()` seems more widely used for this

olsen232 replied:
Done

rcoup commented on generate-list-objects-filter-extensions.sh:
```diff
@@ -0,0 +1,42 @@
+	ARGS=`echo $@ | xargs printf '%s\n'`
+
+	# INVALID_ARGS is those ARGS that don't look like [path/]library.a
+	INVALID_ARGS=`echo "$ARGS" | grep -v '\b[A-Za-z0-9_]\+\.a$'`
```
are there restrictions on library filenames? Seems like `-` at least is surely a common character? Maybe just check it ends with `.a`?

olsen232 replied:
Disallowed simply because it makes linking more complicated: foo.a must define a symbol called `filter_extension_foo`.

Updated the script to fail differently for two different errors - ie, not a path to a library, and name of library is not a valid C symbol.

rcoup commented on Makefile:
```diff
@@ -2261,6 +2278,10 @@ command-list.h: $(wildcard Documentation/git*.txt)
 		$(patsubst %,--exclude-program %,$(EXCLUDED_PROGRAMS)) \
 		command-list.txt >$@+ && mv $@+ $@
 
+list-objects-filter-extensions.c: generate-list-objects-filter-extensions.sh GIT-BUILD-OPTIONS
```
probably need to add stuff to `contrib/buildsystems/CMakeLists.txt` too: `generate-{cmdlist,configlist}.sh` are both mentioned there

And I guess check the MSVC build (see `compat/vcbuild/`) works too

olsen232 replied:
Tried to fix this stuff - hopefully got everything.

rcoup commented on contrib/filter-extensions/rand_cpp/Makefile:
```diff
@@ -0,0 +1,33 @@
+
+ifneq ($(findstring s,$(MAKEFLAGS)),s)
+ifndef V
+	QUIET_CC       = @echo '   ' CC $@;
```
```suggestion
	QUIET_CC       = @echo '   ' CC $@;
	QUIET_CXX       = @echo '   ' CXX $@;
```

olsen232 replied:
Done

rcoup commented on contrib/filter-extensions/rand_cpp/Makefile:
```diff
@@ -0,0 +1,33 @@
+	$(QUIET_AR)$(AR) $(ARFLAGS) $@ $^
+
+rand.o: rand.cpp
+	$(QUIET_CC)$(CC) -c $(ALL_CFLAGS) $(ALL_CPP_FLAGS) $<
```
```suggestion
	$(QUIET_CXX)$(CXX) -c $(ALL_CFLAGS) $(ALL_CXXFLAGS) $<
```

olsen232 replied:
Done

rcoup commented on contrib/filter-extensions/rand_cpp/Makefile:
```diff
@@ -0,0 +1,33 @@
+
+FILTER_STATIC_LIB = rand_cpp.a
+
+ALL_CPP_FLAGS += -std=c++11
```
```suggestion
ALL_CXXFLAGS += -std=c++11
```

olsen232 replied:
Done

rcoup commented on contrib/filter-extensions/rand_cpp/rand.cpp:
```diff
@@ -0,0 +1,105 @@
+	struct rand_context *ctx = static_cast<struct rand_context*>(context);
+
+	struct timespec ended_at;
+	clock_gettime(CLOCK_MONOTONIC, &ended_at);
```
here too

olsen232 replied:
Done

rcoup commented on contrib/filter-extensions/rand_cpp/rand.cpp:
```diff
@@ -0,0 +1,105 @@
+	return 0;
+}
+
+enum list_objects_filter_result rand_filter_object(
```
```suggestion
static enum list_objects_filter_result rand_filter_object(
```

Otherwise it ends up in the global symbol table at link time (ie: will clash with other names)

A namespace (anonymous or other) around everything except `filter_extension_rand_cpp` works too.

olsen232 replied:
Added anonymous namespace around everything

rcoup replied:
should un-static rand_init then

rcoup commented on contrib/filter-extensions/rand_cpp/rand.cpp:
```diff
@@ -0,0 +1,105 @@
+	}
+}
+
+void rand_free(const struct repository *r, void *context) {
```
```suggestion
static void rand_free(const struct repository *r, void *context) {
```

olsen232 replied:
Done - anonymous namespace

rcoup commented on Makefile:
```diff
@@ -1198,6 +1200,7 @@ THIRD_PARTY_SOURCES += sha1dc/%
 GITLIBS = common-main.o $(LIB_FILE) $(XDIFF_LIB)
 EXTLIBS =
 
+
```
stray whitespace

olsen232 replied:
Done

rcoup commented on contrib/filter-extensions/README.md:
```diff
@@ -0,0 +1,141 @@
+$ ./git rev-list refs/heads/master --objects --max-count 1 --filter=extension:rand=3 --filter-print-omitted | grep -c '^~'
+filter-rand: matching 3%
+filter-rand: done: count=4068 (blob=3866 tree=202) matched=117 elapsed=0.005017s rate=810843.1/s average=1.2us
+3749
```
```suggestion
3749  // number of omitted blobs = 3866 - 117
```

olsen232 replied:
Done

rcoup reviewed (commented):
This is looking good IMO. How do we write some tests? 😃

olsen232 committed:
Address rcoup's comments

These commits will eventually be divided up in a more logical way
for git reviewers

rcoup commented on contrib/filter-extensions/rand/rand.c:
```diff
@@ -0,0 +1,103 @@
+	case LOFS_BLOB:
+		ctx->blobCount++;
+
+		if ((rand() % 100) < ctx->percentageMatch) {
```
should add `srand(time(NULL))` somewhere or it won't be random...
